pkg/dia/nft/nftTrade-scrapers/tofunft.go

```diff
@@ -0,0 +1,772 @@
+	if err = s.loadConfig(ctx); err != nil {
+		log.Warnf("unable to load scraper config: %s", err.Error())
+		return err
+	}
+
+	// read state
+	if err = s.loadState(ctx); err != nil {
+		log.Warnf("unable to load scraper state: %s", err.Error())
+		return err
+	}
+
+	log.Infof("fetching tofunft trade transactions from block %d(+%d)", s.state.LastBlockNum, s.conf.BatchSize)
+
+	// fetch trade transactions
+	res, err := utils.EthFilterTXs(ctx, s.tradeScraper.ethConnection, utils.EthTxFilterCriteria{
+		StartBlockNum:      s.state.LastBlockNum,
+		StartTxIndex:       s.state.LastTxIndex,
+		LimitBlocks:        s.conf.BatchSize,
+		BehindHighestBlock: s.conf.FollowDist,
+		EvAddrs:            []common.Address{common.HexToAddress(s.conf.ContractAddr)},
+		Events:             []common.Hash{tofuNFTABI.Events["EvInventoryUpdate"].ID},
+	})
+
+	if err != nil {
+		log.Warnf("unable to filter tofunft trades: %s", err.Error())
+		return err
+	}
+
+	log.Infof("found %d trade(logs: %d) transactions in %d blocks(from %d [tx index offset: %d] to %d, sync: %t[stay behind: -%d]), exploring details...", res.NumTXs, res.NumLogs, res.NumBlocks, s.state.LastBlockNum, s.state.LastTxIndex, res.LastBlockNum, res.Synced, s.conf.FollowDist)
+
+	numTrades := 0
+
+	// process trade transactions
+	for _, tx := range res.TXs {
+		s.state.LastBlockNum = tx.BlockNum
+		s.state.LastTxIndex = tx.TXIndex
+		s.state.LastErr = ""
+		log.Info("current state.ErrCounter: ", s.state.ErrCounter)
+
+		skipped, err := s.processTx(ctx, tx)
+		if err != nil {
+			s.state.ErrCounter++
+
+			if s.state.ErrCounter <= s.conf.MaxRetry {
+				s.state.LastErr = fmt.Sprintf("unable to process trade transaction(%s): %s", tx.TXHash.Hex(), err.Error())
+				log.Error(s.state.LastErr)
+				// store state
+				if err := s.storeState(ctx); err != nil {
+					log.Warnf("unable to store scraper state: %s", err.Error())
+					return err
+				}
+				return err
+			}
+
+			log.Warnf("SKIPPING PERMANENTLY! block: %d, tx index: %d - error: %s", s.state.LastBlockNum, s.state.LastTxIndex, err.Error())
+		}
+
+		if !skipped {
+			numTrades++
+		}
+
+		// reset consecutive error counter
+		s.state.ErrCounter = 0
+
+		// move next
+		s.state.LastTxIndex = tx.TXIndex + 1
+
+		// store state
+		if err := s.storeState(ctx); err != nil {
+			log.Warnf("unable to store scraper state: %s", err.Error())
+			return err
+		}
+	}
+
+	s.state.LastBlockNum = res.LastBlockNum + 1
+	s.state.LastTxIndex = 0
+
+	if err := s.storeState(ctx); err != nil {
+		log.Warnf("unable to store scraper state: %s", err.Error())
+		return err
+	}
+
+	log.Infof("processed %d trades", numTrades)
+
+	return nil
+}
+
+func (s *TofuNFTScraper) processTx(ctx context.Context, tx *utils.EthFilteredTx) (bool, error) {
+	log.Tracef("process tx -> block: %d, tx index: %d, tx hash: %s", tx.BlockNum, tx.TXIndex, tx.TXHash.Hex())
+
+	var ev tofunft.TofunftEvInventoryUpdate
+	err := tofuNFTABI.UnpackIntoInterface(&ev, "EvInventoryUpdate", tx.Logs[0].Data)
```

Comment:
The resulting 0object ev does not hold the data as used later on:
- ev.Raw.BlockNumber is always zero
- ev.Raw.TxHash is always zero